designs/degraded-nodepools.md

```diff
@@ -0,0 +1,119 @@
+
+To improve visibility of these failure modes, this RFC proposes adding a `Degraded` status condition on the Nodepool that indicate to cluster users there may be a problem with a NodePool/NodeClass combination that needs to be investigated and corrected. 
+
+**Note:** The issues that we want to address in the launch path with this status condition are the ones that happen during Nodeclaim launch and registration. Nodeclaim initialization never fails as of today, it can only transition from Unknown to True unless we plan on changing that in the future.   
+
+## 🔑 Introduce a Degraded Status Condition on the NodePool
+
+```
+// ConditionTypeDegraded = "Degraded" condition indicates that a misconfiguration exists that prevents the normal, successful use of a Karpenter resource
+ConditionTypeDegraded = "Degraded"
+```
+We want to introduce a `Degraded` status condition on the Nodepool. This condition can be set to either of Unknown, True or False. The state transition is not unidirectional meaning it can go from True to False and back to True or Unknown. `Degraded` status condition will not be a pre-condition for Nodepool readiness. A degraded Nodepool will continue to be considered for provisioning. We will only deprioritize it during provisioning among the nodepools with similar weight.
+
+### Option 1: In-memory Buffer to store history - Recommended
+
+This option will have an in-memory FIFO buffer, which will grow to a max size of 10 (this can be changed later). This buffer will store data about the success or failure in the launch path and is evaluated by a controller to determine the relative health of the NodePool. This will be an int buffer and a positive means `Degraded: False`, negative means `Degraded: True` and 0 means `Degraded: Unknown`. 
+
+Evaluation conditions -
+
+1. We start with an empty buffer with `Degraded: Unknown`.
+2. There have to be 2 minimum failures in the buffer for `Degraded` to transition to `True` or basically the threshold would be 80%. 
+3. If the buffer starts with a success then `Degraded: False`. 
+4. If Karpenter restarts then we flush the buffer but don't change the existing state of `Degraded` status condition.
+5. If there is an update to a Nodepool/Nodeclass, flush the buffer and set `Degraded: Unknown`.
+6. Since the buffer is FIFO, we remove the oldest launch result when the max buffer size is reached.
+
+See below for example evaluations:
+
+```
+Successful Launch: 1
+Default: 0
+Unsuccessful Launch: -1
+
+[] = 'Degraded: Unknown'
+[-1] = 'Degraded: Unknown'
+[+1] = 'Degraded: False'
+[-1, +1] = 'Degraded: False'
+[-1, -1] = 'Degraded: True'
+[-1, +1, -1] = 'Degraded: True'
+[-1, +1, +1, +1, +1, +1, +1, +1, +1, +1] = 'Degraded: False'
+```
+
+#### Considerations
+
+1. 👎 Heuristics can be wrong and mask failures
```

Comment:
Could you elaborate on what type of failures are being masked? As for it being wrong, I'm wondering if we should only consider degraded unknown or true. Maybe we don't ever transition it to false?